File: tools/release/check_node_direct_release_assets.py
#!/usr/bin/env python3
"""Validate local oliphaunt-node-direct GitHub release assets."""

from __future__ import annotations

import argparse
import hashlib
import sys
import tarfile
import zipfile
from pathlib import Path
from typing import NoReturn

import artifact_targets
import product_metadata


ROOT = Path(__file__).resolve().parents[2]


def fail(message: str) -> NoReturn:
    print(f"check_node_direct_release_assets.py: {message}", file=sys.stderr)
    raise SystemExit(1)


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def checksum_manifest(path: Path) -> dict[str, str]:
    values: dict[str, str] = {}
    for index, raw_line in enumerate(path.read_text(encoding="utf-8").splitlines(), start=1):
        line = raw_line.strip()
        if not line:
            continue
        parts = line.split(maxsplit=1)
        if len(parts) != 2 or len(parts[0]) != 64:
            fail(f"malformed checksum line {index}: {raw_line}")
        values[parts[1].removeprefix("./")] = parts[0].lower()
    return values


def expected_assets(version: str) -> list[str]:
    return artifact_targets.expected_assets("oliphaunt-node-direct", version, surface="github-release")


def expected_addon_assets(version: str) -> list[str]:
    return artifact_targets.expected_assets(
        "oliphaunt-node-direct",
        version,
        surface="github-release",
        kinds=["node-direct-addon"],
    )


def addon_targets_by_asset(version: str) -> dict[str, artifact_targets.ArtifactTarget]:
    return {
        target.asset_name(version): target
        for target in artifact_targets.artifact_targets(
            product="oliphaunt-node-direct",
            surface="github-release",
            published_only=True,
        )
        if target.kind == "node-direct-addon"
    }


def validate_tar_archive(path: Path, member_name: str) -> None:
    with tarfile.open(path, "r:gz") as archive:
        names = set(archive.getnames())
        if member_name not in names:
            fail(f"{path.name} is missing {member_name}")
        member = archive.getmember(member_name)
        if not member.isfile():
            fail(f"{path.name} {member_name} is not a regular file")
        if member.size == 0:
            fail(f"{path.name} {member_name} is empty")


def validate_zip_archive(path: Path, member_name: str) -> None:
    with zipfile.ZipFile(path) as archive:
        names = set(archive.namelist())
        if member_name not in names:
            fail(f"{path.name} is missing {member_name}")
        member = archive.getinfo(member_name)
        if member.is_dir():
            fail(f"{path.name} {member_name} is not a regular file")
        if member.file_size == 0:
            fail(f"{path.name} {member_name} is empty")


def validate_addon_archive(path: Path, target: artifact_targets.ArtifactTarget) -> None:
    member_name = target.library_relative_path
    if member_name is None:
        fail(f"{target.id} is missing library_relative_path")
    if path.name.endswith(".tar.gz"):
        validate_tar_archive(path, member_name)
    elif path.suffix == ".zip":
        validate_zip_archive(path, member_name)
    else:
        fail(f"{path.name} has unsupported Node direct archive extension")


def validate(asset_dir: Path, allow_partial: bool = False) -> None:
    version = product_metadata.read_current_version("oliphaunt-node-direct")
    required_assets = expected_assets(version)
    addon_targets = addon_targets_by_asset(version)
    missing = [asset for asset in required_assets if not (asset_dir / asset).is_file()]
    if missing:
        if not allow_partial:
            fail("missing oliphaunt-node-direct release asset(s): " + ", ".join(missing))
        present_addons = [asset for asset in expected_addon_assets(version) if (asset_dir / asset).is_file()]
        if not present_addons:
            fail("partial oliphaunt-node-direct release asset validation requires at least one addon asset")

    checksum_asset = asset_dir / f"oliphaunt-node-direct-{version}-release-assets.sha256"
    if not checksum_asset.is_file():
        fail(f"missing checksum manifest: {checksum_asset.name}")
    checksums = checksum_manifest(checksum_asset)
    for asset in required_assets:
        if allow_partial and not (asset_dir / asset).is_file():
            continue
        if asset == checksum_asset.name:
            continue
        expected_digest = checksums.get(asset)
        if expected_digest is None:
            fail(f"{checksum_asset.name} does not cover {asset}")
        actual = sha256(asset_dir / asset)
        if actual != expected_digest:
            fail(f"checksum mismatch for {asset}: expected {expected_digest}, got {actual}")
    for asset in expected_addon_assets(version):
        if allow_partial and not (asset_dir / asset).is_file():
            continue
        target = addon_targets.get(asset)
        if target is None:
            fail(f"no artifact target metadata found for {asset}")
        validate_addon_archive(asset_dir / asset, target)


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--asset-dir",
        default=str(ROOT / "target/oliphaunt-node-direct/release-assets"),
        help="directory containing oliphaunt-node-direct release assets",
    )
    parser.add_argument(
        "--allow-partial",
        action="store_true",
        help="validate the Node direct assets present in asset-dir without requiring every published target",
    )
    args = parser.parse_args(argv)
    validate(Path(args.asset_dir).resolve(), allow_partial=args.allow_partial)
    print(f"oliphaunt-node-direct release assets validated: {Path(args.asset_dir).resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: tools/release/check-node-direct-release-assets.mjs
#!/usr/bin/env bun
import path from "node:path";
import { fileURLToPath } from "node:url";

import {
  assertFileExists,
  checksumManifest,
  readArchiveEntries,
  sha256,
} from "./release-asset-validation.mjs";
import {
  ROOT,
  artifactTargets,
  compareText,
  currentProductVersion,
  expectedAssets,
  fail,
} from "./release-artifact-targets.mjs";
import { inspectPlatformBinaryEntries } from "./platform-binary-contract.mjs";
import { readPortableArchiveEntries } from "./portable-archive.mjs";
import {
  assertReleaseNoticesInEntries,
  releaseProfilePackageLicense,
} from "./release-notices.mjs";

const PREFIX = "check-node-direct-release-assets.mjs";
const PRODUCT = "oliphaunt-node-direct";
const KIND = "node-direct-addon";
const NOTICE_OPTIONS = Object.freeze({ profile: "source-sdk" });
const NOTICE_MEMBERS = Object.freeze(["LICENSE", "THIRD_PARTY_NOTICES.md"]);
const PACKAGE_LICENSE = releaseProfilePackageLicense("source-sdk").spdx;

function parseArgs(argv) {
  const args = {
    assetDir: path.join(ROOT, "target/oliphaunt-node-direct/release-assets"),
    allowPartial: false,
    npmPackages: [],
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--asset-dir") {
      const value = argv[index + 1];
      if (!value) {
        fail(PREFIX, "--asset-dir requires a value");
      }
      args.assetDir = path.resolve(value);
      index += 1;
    } else if (arg === "--allow-partial") {
      args.allowPartial = true;
    } else if (arg === "--npm-package") {
      const value = argv[index + 1];
      if (!value) {
        fail(PREFIX, "--npm-package requires a value");
      }
      args.npmPackages.push(path.resolve(value));
      index += 1;
    } else {
      fail(PREFIX, `unknown argument ${arg}`);
    }
  }
  return args;
}

export function assertNodeDirectReleaseNoticeEntries(entries, { prefix = "", label = "Node direct archive" } = {}) {
  return assertReleaseNoticesInEntries(entries, {
    ...NOTICE_OPTIONS,
    prefix,
    label,
  });
}

function archiveJson(entries, member, label) {
  const entry = entries.get(member);
  if (!entry?.isFile || entry.isSymbolicLink) {
    throw new Error(`${label} is missing regular member ${member}`);
  }
  if ((entry.mode & 0o777) !== 0o644) {
    throw new Error(`${label} member ${member} must have mode 0644`);
  }
  try {
    return JSON.parse(Buffer.from(entry.data()).toString("utf8"));
  } catch (cause) {
    throw new Error(`${label} member ${member} must contain valid JSON: ${cause.message}`);
  }
}

export function assertNodeDirectNpmArchive(file, targets, version) {
  const label = path.basename(file);
  const entries = readArchiveEntriesForNotices(file, label);
  assertNodeDirectReleaseNoticeEntries(entries, { prefix: "package", label });
  const manifest = archiveJson(entries, "package/package.json", label);
  const target = targets.find((candidate) => candidate.npmPackage === manifest.name);
  if (!target) {
    throw new Error(`${label} package name is not a published Node direct carrier: ${JSON.stringify(manifest.name)}`);
  }
  if (manifest.version !== version) {
    throw new Error(`${label} package version must be ${version}, got ${JSON.stringify(manifest.version)}`);
  }
  if (manifest.license !== PACKAGE_LICENSE) {
    throw new Error(`${label} package license must be ${PACKAGE_LICENSE}, got ${JSON.stringify(manifest.license)}`);
  }
  if (manifest.oliphaunt?.target !== target.target) {
    throw new Error(
      `${label} package target must be ${target.target}, got ${JSON.stringify(manifest.oliphaunt?.target)}`,
    );
  }
  if (!Array.isArray(manifest.files)) {
    throw new Error(`${label} package.json must declare an npm files allowlist`);
  }
  if (
    manifest.files.some((member) => typeof member !== "string" || member.length === 0)
    || new Set(manifest.files).size !== manifest.files.length
  ) {
    throw new Error(`${label} package.json npm files allowlist must contain unique non-empty strings`);
  }
  for (const member of NOTICE_MEMBERS) {
    if (!manifest.files.includes(member)) {
      throw new Error(`${label} package.json npm files allowlist must include ${member}`);
    }
  }
  const prebuild = entries.get("package/prebuilds/oliphaunt_node.node");
  if (!prebuild?.isFile || prebuild.isSymbolicLink || prebuild.size === 0) {
    throw new Error(`${label} is missing a non-empty regular package/prebuilds/oliphaunt_node.node`);
  }
  return manifest;
}

function readArchiveEntriesForNotices(file, label) {
  try {
    return readPortableArchiveEntries(file);
  } catch (error) {
    throw new Error(`${label} is not a valid portable archive: ${error.message}`);
  }
}

async function validateArchive(file, target) {
  const entries = await readArchiveEntries(file, fail, PREFIX, "Node direct");
  try {
    assertNodeDirectReleaseNoticeEntries(entries, { label: path.basename(file) });
  } catch (error) {
    fail(PREFIX, error.message);
  }
  const memberName = target.libraryRelativePath;
  if (!entries.has(memberName)) {
    fail(PREFIX, `${path.basename(file)} is missing ${memberName}`);
  }
  const member = entries.get(memberName);
  if (!member.isFile) {
    fail(PREFIX, `${path.basename(file)} ${memberName} is not a regular file`);
  }
  if (member.size === 0) {
    fail(PREFIX, `${path.basename(file)} ${memberName} is empty`);
  }
  inspectPlatformBinaryEntries(
    [...entries].map(([name, entry]) => ({ name, ...entry })),
    { target: target.target, rootLabel: path.basename(file) },
  );
}

async function main(argv) {
  const args = parseArgs(argv);
  const version = await currentProductVersion(PRODUCT, PREFIX);
  const requiredAssets = expectedAssets(PRODUCT, KIND, version, PREFIX);
  const targets = artifactTargets(PRODUCT, KIND, PREFIX);
  const targetsByAsset = new Map(targets.map((target) => [target.asset.replaceAll("{version}", version), target]));
  const missing = [];
  for (const asset of requiredAssets) {
    if (!(await assertFileExists(path.join(args.assetDir, asset)))) {
      missing.push(asset);
    }
  }
  if (missing.length > 0) {
    if (!args.allowPartial) {
      fail(PREFIX, `missing oliphaunt-node-direct release asset(s): ${missing.join(", ")}`);
    }
    let presentAddons = 0;
    for (const target of targets) {
      if (await assertFileExists(path.join(args.assetDir, target.asset.replaceAll("{version}", version)))) {
        presentAddons += 1;
      }
    }
    if (presentAddons === 0) {
      fail(PREFIX, "partial oliphaunt-node-direct release asset validation requires at least one addon asset");
    }
  }

  const checksumAsset = `oliphaunt-node-direct-${version}-release-assets.sha256`;
  const checksumPath = path.join(args.assetDir, checksumAsset);
  if (!(await assertFileExists(checksumPath))) {
    fail(PREFIX, `missing checksum manifest: ${checksumAsset}`);
  }
  const checksums = await checksumManifest(checksumPath, fail, PREFIX);
  for (const asset of requiredAssets.sort(compareText)) {
    const assetPath = path.join(args.assetDir, asset);
    if (args.allowPartial && !(await assertFileExists(assetPath))) {
      continue;
    }
    if (asset === checksumAsset) {
      continue;
    }
    const expected = checksums.get(asset);
    if (!expected) {
      fail(PREFIX, `${checksumAsset} does not cover ${asset}`);
    }
    const actual = await sha256(assetPath);
    if (actual !== expected) {
      fail(PREFIX, `checksum mismatch for ${asset}: expected ${expected}, got ${actual}`);
    }
  }
  for (const [asset, target] of targetsByAsset) {
    const assetPath = path.join(args.assetDir, asset);
    if (args.allowPartial && !(await assertFileExists(assetPath))) {
      continue;
    }
    await validateArchive(assetPath, target);
  }
  for (const npmPackage of args.npmPackages) {
    try {
      assertNodeDirectNpmArchive(npmPackage, targets, version);
    } catch (error) {
      fail(PREFIX, error.message);
    }
  }
  console.log(`oliphaunt-node-direct release assets validated: ${args.assetDir}`);
}

const invoked = process.argv[1] ? path.resolve(process.argv[1]) : "";
if (invoked === fileURLToPath(import.meta.url)) {
  await main(Bun.argv.slice(2));
}
